### PR TITLE
Master iap lead enrich v2 fle

### PR DESCRIPTION
**Pull request by fleodoo**
Purpose:
Allow the user to enrich a lead based on the domain of the email adress of the client.
Following fields will be populated if found using clearbit and if they are not yet populated:
- description
- partner_name
- reveal_id
- website
- street, street2, zip, city, country_id, state_id
- phone (using the first number found in clearbit)
- mobile (using the second number found in clearbit, else the same as phone)
A message will also be logged in the chatter.
If no data is found using clearbit no credit is used and a message will be logged,
If the user has no credits he will be redirected to buy credits for the service.
It is possible to enrich  using the 'enrich' action in listview or with a cron that will run every hour.

This commit includes:
- a new action in listview
- a new message tempalate to log when no data is found in clearbit
- a new message tempalate to log when the email is not valid
- a new message tempalate to log when no credit is left
- a new cron
- additions to crm_lead to be able to enrich

TaskId: 193185

--
I confirm I have signed the CLA and read the PR guidelines at www.odoo.com/submit-pr

### CODE REVIEW

File: addons/crm_iap_lead_enrich/__manifest__.py [hunk 1]
```diff
@@ -0,0 +1,17 @@
+# -*- coding: utf-8 -*-
+# Part of Odoo. See LICENSE file for full copyright and licensing details.
+
+{
+    'name': 'Lead Enrichment',
+    'summary': 'Enrich Leads/Opportunities using email adrress domain and clearbit',
```
Comment (pro-odoo) [inline addons/crm_iap_lead_enrich/__manifest__.py]: 'Enrich Leads/Opportunities using email address domain'

File: addons/crm_iap_lead_enrich/models/crm_iap_lead_enrich.py [hunk 2]
```diff
@@ -0,0 +1,124 @@
...
+        for record in self:
+            if record.probability == 0:
+                continue
+            email_domain = record.partner_address_email or record.email_from
+            if email_domain:
+                request[record.id] = email_domain.split('@')[1]
+            else:
+                record.message_post_with_view('crm_iap_lead_enrich.lead_enrich_message_no_mail', subtype_id=self.env.ref('mail.mt_note').id)        
+        try:
+            response_clearbit = record._make_request(request)
```
Comment (pro-odoo) [inline addons/crm_iap_lead_enrich/models/crm_iap_lead_enrich.py]: `self._make_request(request)`

File: addons/crm_iap_lead_enrich/models/crm_iap_lead_enrich.py [hunk 3]
```diff
@@ -0,0 +1,124 @@
...
+            self._enrich_leads_from_response(response_clearbit)
+        except InsufficientCreditError:
+            for lead in self:
```
Comment (pro-odoo) [inline addons/crm_iap_lead_enrich/models/crm_iap_lead_enrich.py]: Be consistent, either you use `for record in self` or `for lead in self`, but keep the same name (line 15 and here)

File: addons/crm_iap_lead_enrich/models/crm_iap_lead_enrich.py [hunk 4]
```diff
@@ -0,0 +1,124 @@
...
+                data = {
+                    'message_title': _("Lead enriched based on email address :D"),
```
Comment (pro-odoo) [inline addons/crm_iap_lead_enrich/models/crm_iap_lead_enrich.py]: No ':)' :)

File: addons/crm_iap_lead_enrich/models/crm_iap_lead_enrich.py [hunk 5]
```diff
@@ -0,0 +1,124 @@
...
+                    'url' : self.env['iap.account'].get_account_url(),
+                }
+                lead.message_post_with_view('crm_iap_lead_enrich.lead_enrich_message_no_credit', values=data, subtype_id=self.env.ref('mail.mt_note').id)
+
+    @api.multi
+    def lead_enrich_mail_server_action(self):
```
Comment (pro-odoo) [inline addons/crm_iap_lead_enrich/models/crm_iap_lead_enrich.py]: Why this method, and not simply call `lead_enrich_mail()' ?

File: addons/crm_iap_lead_enrich/models/crm_iap_lead_enrich.py [hunk 6]
```diff
@@ -0,0 +1,124 @@
...
+        self.lead_enrich_mail()
+    
+    @api.multi
```
Comment (pro-odoo) [inline addons/crm_iap_lead_enrich/models/crm_iap_lead_enrich.py]: model

File: addons/iap/static/src/js/crash_manager.js [hunk 7]
```diff
@@ -20,7 +20,7 @@ CrashManager.include({
      */
     _getButtonMessage: function (isTrial){
         var isEnterprise = _.last(odoo.session_info.server_version_info) === 'e';
-        return isTrial && isEnterprise ? _t('Start a Trial at Odoo') : _t('Buy credits at Odoo');
+        return isTrial && isEnterprise ? _t('Start a Trial at Odoo') : _t('Buy credits');
```
Comment (pro-odoo) [inline addons/iap/static/src/js/crash_manager.js]: Move it to another commit

**Review by pro-odoo — CHANGES_REQUESTED** [review]
Some little remarks

File: addons/crm_iap_lead_enrich/models/crm_iap_lead_enrich.py [hunk 8]
```diff
@@ -0,0 +1,120 @@
...
+            for record in self:
+                data = {
+                    'message_title': _("Lead enriched based on email address"),
+                    'url' : self.env['iap.account'].get_account_url(),
+                }
+                record.message_post_with_view('crm_iap_lead_enrich.lead_enrich_message_no_credit', values=data, subtype_id=self.env.ref('mail.mt_note').id)
+    
+    @api.model
+    def enrich_with_cron(self):
+        leads = self.search([('enriched_lead', '=', False),('reveal_id', '=', False)])
```
Comment (pro-odoo) [inline addons/crm_iap_lead_enrich/models/crm_iap_lead_enrich.py]: Hmmm, make no sense, all old leads, leads in won stage, ... will be processed for nothing... We should change the condition to something more restrictive.

Comment (fleodoo) [reply addons/crm_iap_lead_enrich/models/crm_iap_lead_enrich.py]: who this be restrictive enough? 
leads = self.search([('enriched_lead', '=', False),('reveal_id', '=', False),('active','=',True),('won_status','=','pending')])

Comment (pro-odoo) [reply addons/crm_iap_lead_enrich/models/crm_iap_lead_enrich.py]: When you search on records, active=True is implicit, you do not need to add this condition.
For the condition, you can have a look in the task :)

File: addons/crm_iap_lead_enrich/models/__init__.py [hunk 9]
```diff
@@ -0,0 +1,4 @@
+# -*- coding: utf-8 -*-
```
Comment (pro-odoo) [inline addons/crm_iap_lead_enrich/models/__init__.py]: Add these two lines in all the python files

**Review by pro-odoo — CHANGES_REQUESTED** [review]

File: addons/crm_iap_lead_enrich/models/crm_iap_lead_enrich.py [hunk 10]
```diff
@@ -0,0 +1,150 @@
...
+                        'zip': record.zip or zip_code,
+                        'country_id': record.country_id.id or country_id,
+                        'state_id': record.state_id.id or state_id,
+                        'enriched_lead': True,
+                    })
+                    record._send_message(data)
+                else:
+                    record._send_message()
+    @api.model
+    def notify_no_more_credit(self, service_name, model_name, notification_parameter):
```
Comment (tde-banana-odoo) [inline addons/crm_iap_lead_enrich/models/crm_iap_lead_enrich.py]: Well I don't know what are the real specifications but I don't think you want to spam every people that once created leads. What are the people to warn ?

File: addons/crm/__init__.py [hunk 11]
```diff
@@ -8,3 +8,8 @@
 
 from odoo import api, SUPERUSER_ID
 
+def _auto_install_iap_lead_startup(cr, registry):
```
Comment (tde-banana-odoo) [inline addons/crm/__init__.py]: What is the purpose of this code ?

File: addons/crm_iap_lead_enrich/data/mail_template.xml [hunk 12]
```diff
@@ -0,0 +1,22 @@
+<?xml version="1.0" encoding="utf-8"?>
+<odoo>
+    <data noupdate="1">
+        <record id="lead_enrichment_no_credits" model="mail.template">
```
Comment (tde-banana-odoo) [inline addons/crm_iap_lead_enrich/data/mail_template.xml]: This template is quite dummy and does not really work as it is defined. I would instead make a template on res.users as you are trying to contact users, with partner_to correctly set.

File: addons/crm_iap_lead_enrich/models/crm_iap_lead_enrich.py [hunk 13]
```diff
@@ -0,0 +1,150 @@
...
+                self.message_post_with_view('crm_iap_lead_enrich.lead_enrich_message_no_credit', values=data, subtype_id=self.env.ref('mail.mt_note').id)
+
+    @api.model
+    def enrich_with_cron(self):
+        timeDelta = fields.datetime.now() - datetime.timedelta(hours=1)
+        leads = self.search([('enriched_lead', '=', False),('reveal_id', '=', False),('probability','!=','0'),('probability','!=','100'),('create_date','>',timeDelta)])
+        leads.lead_enrich_mail(True)
+
+    @api.model
+    def _make_request(self, domains):
```
Comment (tde-banana-odoo) [inline addons/crm_iap_lead_enrich/models/crm_iap_lead_enrich.py]: I would make a small crm_lead_iap.py file with method to contact iap, like Pro done in sms (see sms.sms_api.py) . It allows to separate real business code from iap code.

File: addons/crm_iap_lead_enrich/models/crm_iap_lead_enrich.py [hunk 14]
```diff
@@ -0,0 +1,150 @@
...
+from odoo import fields, models, api, _
+from odoo.exceptions import UserError
+from odoo.addons.iap import jsonrpc,InsufficientCreditError
+
+DEFAULT_ENDPOINT = 'https://iap-services.odoo.com'
+
+class Lead(models.Model):
+    _inherit = 'crm.lead'
+
+    enriched_lead = fields.Boolean(string='Lead enriched based on email', default=False)
```
Comment (tde-banana-odoo) [inline addons/crm_iap_lead_enrich/models/crm_iap_lead_enrich.py]: As False is already the defualt value of a boolean field there is no need to force it.

File: addons/crm_iap_lead_enrich/data/message_template.xml [hunk 15]
```diff
@@ -0,0 +1,26 @@
+<?xml version="1.0" encoding="utf-8"?>
```
Comment (tde-banana-odoo) [inline addons/crm_iap_lead_enrich/data/message_template.xml]: You can rename to mail_data.xml like in other modules, just for naming convention.

File: addons/crm_iap_lead_enrich/data/message_template.xml [hunk 16]
```diff
@@ -0,0 +1,26 @@
...
+
+<odoo>
+    <template id="lead_enrich_message_not_found">
```
Comment (tde-banana-odoo) [inline addons/crm_iap_lead_enrich/data/message_template.xml]: I would rename to something like 'mail_message_lead_enrich_notfound'. Purpose is to name space: used as core of mail.message, linked to lead, then suffix to understand its purpose.

File: addons/crm_iap_lead_enrich/data/message_template.xml [hunk 17]
```diff
@@ -0,0 +1,26 @@
...
+        <p>Lead Enrichment based on email address</p>
+        <div style="background-color:#ffffff;padding:15px;">
+            <span> No company data found based on the email address or email address is one of an email provider. No credit was consumed. </span>
+        </div>
+    </template>
+
+    <template id="lead_enrich_message_no_mail">
```
Comment (tde-banana-odoo) [inline addons/crm_iap_lead_enrich/data/message_template.xml]: For example here: mail_message_lead_enrich_no_email

File: addons/crm_iap_lead_enrich/data/message_template.xml [hunk 18]
```diff
@@ -0,0 +1,26 @@
...
+        <p>Lead Enrichment based on email address</p>
+        <div style="background-color:#ffffff;padding:15px;">
+            <span>Enrichment could not be done as no email address was provided.</span>
+        </div>
+    </template>
+
+    <template id="lead_enrich_message_no_credit">
```
Comment (tde-banana-odoo) [inline addons/crm_iap_lead_enrich/data/message_template.xml]: For example here: mail_message_lead_enrich_no_credit

File: addons/crm_iap_lead_enrich/models/crm_iap_lead_enrich.py [hunk 19]
```diff
@@ -0,0 +1,150 @@
...
+                response_clearbit = self._make_request(domains)
+                self._enrich_leads_from_response(response_clearbit)
+                self.env['ir.config_parameter'].sudo().set_param('lead_enrich.already_notified', False)
+
+            except InsufficientCreditError:
+                self.notify_no_more_credit('reveal', self._name, 'lead_enrich.already_notified')
+                if cron:
+                    return
+                data = {
+                    'message_title': _("Lead enriched based on email address"),
```
Comment (tde-banana-odoo) [inline addons/crm_iap_lead_enrich/models/crm_iap_lead_enrich.py]: No need to use a variable for a string that you could directly embed in lead_enrich_message_no_credit template.

File: addons/crm_iap_lead_enrich/models/crm_iap_lead_enrich.py [hunk 20]
```diff
@@ -0,0 +1,150 @@
...
+            log_data.update({
+                'timezone': timezone.replace('_', ' ').title(),
+                'timezone_url': company_data['timezone_url'],
+            })
+        return log_data
+
+    @api.model
+    def _send_message(self, company_data=False):
+        if company_data:
+            messages_to_post = self._create_message_data(company_data)
```
Comment (tde-banana-odoo) [inline addons/crm_iap_lead_enrich/models/crm_iap_lead_enrich.py]: Woudln't it be possible to give the created lead to render in lead_message_template (well in a copy of it, upgraded to work on a record) ? That way you don't have to go through various marshmalling methods that just format data in a way or another. Normally everything you store should be present in the lead and therefore you may give it directly to a qweb template. Do you think this would work ?

File: addons/crm_iap_lead_enrich/models/crm_iap_lead_enrich.py [hunk 21]
```diff
@@ -0,0 +1,150 @@
...
+            self.message_post_with_view('crm_iap_lead.lead_message_template', values=messages_to_post, subtype_id=self.env.ref('mail.mt_note').id)
+        else:
+            self.message_post_with_view('crm_iap_lead_enrich.lead_enrich_message_not_found', subtype_id=self.env.ref('mail.mt_note').id)
```
Comment (tde-banana-odoo) [inline addons/crm_iap_lead_enrich/models/crm_iap_lead_enrich.py]: Well that's strange that _send_message has to decide whether to send one message or another. I would totally get rid of _send_message and post directly in the caller, using the right template. Otherwise you have methods that behave differently depending on some data and that is quite hard to read.

File: addons/crm_iap_lead_enrich/models/crm_iap_lead_enrich.py [hunk 22]
```diff
@@ -0,0 +1,150 @@
...
+
+DEFAULT_ENDPOINT = 'https://iap-services.odoo.com'
+
+class Lead(models.Model):
+    _inherit = 'crm.lead'
+
+    enriched_lead = fields.Boolean(string='Lead enriched based on email', default=False)
+
+    @api.model
+    def lead_enrich_mail(self, cron=False):
```
Comment (tde-banana-odoo) [inline addons/crm_iap_lead_enrich/models/crm_iap_lead_enrich.py]: I would make this method private to avoid people calling it from outside of odoo.

File: addons/crm_iap_lead_enrich/models/crm_iap_lead_enrich.py [hunk 23]
```diff
@@ -0,0 +1,150 @@
...
+                if cron:
+                    return
+                data = {
+                    'message_title': _("Lead enriched based on email address"),
+                    'url' : self.env['iap.account'].get_credits_url('reveal'),
+                }
+                self.message_post_with_view('crm_iap_lead_enrich.lead_enrich_message_no_credit', values=data, subtype_id=self.env.ref('mail.mt_note').id)
+
+    @api.model
+    def enrich_with_cron(self):
```
Comment (tde-banana-odoo) [inline addons/crm_iap_lead_enrich/models/crm_iap_lead_enrich.py]: I would make this method private to avoid people calling it from outside of odoo.

**Review by tde-banana-odoo — CHANGES_REQUESTED** [review]
First technical review: global code architecure

File: addons/crm/__init__.py (hunk 11 above)
Comment (fleodoo) [reply addons/crm/__init__.py]: Fabien asked that when we install the crm, by default lead enrichment is active and installed.
I first tried using autoinstall. But then it was not possible to uninstall the module anymore. This because
crm_iap_lead_enrichment depends on crm_iap_lead who depends on crm. I had put autoinstall on crm_iap_lead and crm_iap_lead_enrichment. But then when uninstalling crm_iap_lead_enrichment it would reinstall immediately as  crm_iap_lead was still present and can't be removed. Is there another/better way?

Comment (tde-banana-odoo) [reply addons/crm/__init__.py]: Remove all that crap and make them not auto install. This can be done in custom on saas if necessary.

File: addons/crm_iap_lead_enrich/models/crm_iap_lead_enrich.py (hunk 10 above)
Comment (fleodoo) [reply addons/crm_iap_lead_enrich/models/crm_iap_lead_enrich.py]: indeed, I reread the specs, we just need to send a mail to the person in charge of the IAP service. I changed it.

File: addons/crm_iap_lead_enrich/data/mail_template.xml (hunk 12 above)
Comment (fleodoo) [reply addons/crm_iap_lead_enrich/data/mail_template.xml]: As I'm am not contacting users anymore but the admin in charge of the IAP service, I think I should still use iap.model_iap_account. But I now set <field name="email_to">${object.create_uid.email}</field> directly in the template.
and not in the values passed to the template.
